### ISSUE TITLE
Mongoose creates invalid $set objects with conflicts, when updating documents

### ISSUE THREAD

**Issue by plegner** [Automattic/mongoose#10677]
**What is the current behavior?**
Mongoose creates invalid $set objects that contain conflicts when updating documents. It happens if a Mongoose map contains objects, and the keys in the maps contain multiple similar strings like `abc` and `abc-1`.

**If the current behavior is a bug, please provide the steps to reproduce.**
```js
const mongoose = require('mongoose');
mongoose.set('debug', true);
mongoose.connect('mongodb://localhost:27017/', {});

const schema = new mongoose.Schema({
  values: {
    type: Map,
    of: {entries: String},
    default: {}
  }
});

const Model = mongoose.model('test', schema, 'test');

(async () => {
  let saved = new Model({});
  await saved.save();

  let document = await Model.findById(saved.id);
  document.values.set('abc', {entries: 'a'});
  document.values.set('abc-1', {entries: 'b'});
  await document.save();
})();
```

**What is the expected behavior?**
The code above should work without issues. Instead, the following output is observed:
```
Mongoose: test.updateOne({ _id: new ObjectId("...") }, { '$set': { 'values.abc': { entries: 'a', _id: new ObjectId("...") }, 'values.abc-1': { entries: 'b', _id: new ObjectId("...") }, 'values.abc.entries': 'a' }}, { session: null })
(node:3884) UnhandledPromiseRejectionWarning: MongoServerError: Updating the path 'values.abc.entries' would create a conflict at 'values.abc'
```
Notice how the `values.abc` entry is set twice, causing MongoDB to throw an error.

Based on some quick testing, it seems to happen if two map keys have the same prefix (e.g. `abc`), and one of them contains the `-` symbol).

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Mongoose v6.0.4
Node v14.16.0

**Comment by IslandRhythms** [comment]
What version of mongodb?

**Comment by plegner** [comment]
> What version of mongodb?

4.2.7

**Comment by IslandRhythms** [comment]
that version isn't available in the npm registry

**Comment by plegner** [comment]
The MongoDB version should make no difference for this bug, but you can download all the latest versions here:
https://www.mongodb.com/download-center/community/releases

**Comment by IslandRhythms** [comment]
What version do you have from npm.

**Comment by IslandRhythms** [comment]
I'm getting a MongoAPI Error: database names cannot contain the character `'.'`

**Comment by plegner** [comment]
The database name on line 3 is only a placeholder, you need to replace the `...` with an actual database name you want to use

**Comment by vkarpov15** [comment]
Should be fixed by #10709

**Comment by plegner** [comment]
Thanks! 🎉
